### ISSUE TITLE
Samples directory very slow for single departement

### ISSUE THREAD

**Issue by MartinWahnschaffe** [hzi-braunschweig/SORMAS-Project#9753]
### Problem Description
Take about 3 minutes by report of german health department, when using surveillance supervisor, contact supervisor or physician role.

Based on the slow query logs we received, it's clearly the sample directory (mean time of 14 calls is 107453 ms -> almost 2 minutes + another 19 seconds for the related count query).

The problem is not reproducible on test-performance.

I have created an explain for the query based on a local test database with 15k samples and 850 pathogen tests. 
https://explain.dalibo.com/plan/yvJ

The main problem seems to be what we already discovered in #8946: A very big query with a lot of data to be retrieved. Limiting the result does not make a difference because the results are sorted (+ distinct).

It's not clear to me why this would be different for different user roles, though.

### Proposed Solution

Analyse the reported problem in depth and see where the time is lost.

### Additional Information

Findings to be solved:
 - #9779

**Comment by StefanKock** [comment]
> It's not clear to me why this would be different for different user roles, though.

1. I assume the `createUserFilter` for the regional user causes way more delay because of more joins than for `JurisdictionLevel.NATION`. Basically there are three one-to-many joins from Case to Contact that lets the dataset grow exponentially if cases have multiple contacts. Also we know the the more joins there are, there is basically delays for each following join if the dataset at hand gets too big.
2. The join to `LEFT OUTER JOIN Person person18_ ON eventparti11_.person_id=person18_.id` takes about 2s of the 3,5s in the example given by Martin. This maybe the more important hint.

<details>
<summary>SQL comparison regional user vs. no user filter</summary>

```sql
-- SQL of regional user
-- ...
FROM samples sample0_
LEFT OUTER JOIN cases case1_ ON sample0_.associatedCase_id=case1_.id
LEFT OUTER JOIN District district2_ ON case1_.district_id=district2_.id
LEFT OUTER JOIN Person person3_ ON case1_.person_id=person3_.id
LEFT OUTER JOIN contact contacts4_ ON case1_.id=contacts4_.caze_id  -- additional
LEFT OUTER JOIN contact contact5_ ON sample0_.associatedContact_id=contact5_.id
LEFT OUTER JOIN District district6_ ON contact5_.district_id=district6_.id
LEFT OUTER JOIN cases case7_ ON contact5_.caze_id=case7_.id
LEFT OUTER JOIN District district8_ ON case7_.district_id=district8_.id
LEFT OUTER JOIN contact contacts9_ ON case7_.id=contacts9_.caze_id  -- additional
LEFT OUTER JOIN Person person10_ ON contact5_.person_id=person10_.id
LEFT OUTER JOIN EventParticipant eventparti11_ ON sample0_.associatedEventParticipant_id=eventparti11_.id
LEFT OUTER JOIN EVENTS event12_ ON eventparti11_.event_id=event12_.id
LEFT OUTER JOIN LOCATION location13_ ON event12_.eventLocation_id=location13_.id
LEFT OUTER JOIN District district14_ ON location13_.district_id=district14_.id
LEFT OUTER JOIN EventParticipant eventperso15_ ON event12_.id=eventperso15_.event_id  -- additional
LEFT OUTER JOIN cases case16_ ON eventperso15_.resultingCase_id=case16_.id  -- additional
LEFT OUTER JOIN contact contacts17_ ON case16_.id=contacts17_.caze_id  -- additional
LEFT OUTER JOIN Person person18_ ON eventparti11_.person_id=person18_.id
LEFT OUTER JOIN Facility facility19_ ON sample0_.lab_id=facility19_.id
LEFT OUTER JOIN samples sample20_ ON sample0_.referredTo_id=sample20_.id
WHERE (case1_.region_id=1
       OR case1_.responsibleRegion_id=1
       OR contacts4_.reportingUser_id=19
       OR contacts4_.contactOfficer_id=19
       OR contacts4_.region_id=1
       OR case1_.reportingUser_id=19
       OR case1_.surveillanceOfficer_id=19
       OR case1_.caseOfficer_id=19
       OR contact5_.reportingUser_id=19
       OR contact5_.contactOfficer_id=19
       OR contact5_.region_id=1
       OR case7_.region_id=1
       OR case7_.responsibleRegion_id=1
       OR contacts9_.reportingUser_id=19
       OR contacts9_.contactOfficer_id=19
       OR contacts9_.region_id=1
       OR case7_.reportingUser_id=19
       OR case7_.surveillanceOfficer_id=19
       OR case7_.caseOfficer_id=19
       OR location13_.region_id=1
       OR case16_.region_id=1
       OR case16_.responsibleRegion_id=1
       OR contacts17_.reportingUser_id=19
       OR contacts17_.contactOfficer_id=19
       OR contacts17_.region_id=1
       OR case16_.reportingUser_id=19
       OR case16_.surveillanceOfficer_id=19
       OR case16_.caseOfficer_id=19
       OR eventperso15_.region_id=1
       OR location13_.region_id=1
       OR event12_.reportingUser_id=19
       OR event12_.responsibleUser_id=19)
  AND (case1_.archived=FALSE
       OR case1_.archived IS NULL)
  AND sample0_.deleted=FALSE
ORDER BY sample0_.sampleDateTime DESC
```

```sql
-- SampleFacadeEjbTest.testGetIndexList -> no user = less filtering = less joins
-- ...
FROM samples sample0_
LEFT OUTER JOIN cases case1_ ON sample0_.associatedCase_id=case1_.id
LEFT OUTER JOIN District district2_ ON case1_.district_id=district2_.id
LEFT OUTER JOIN Person person3_ ON case1_.person_id=person3_.id
LEFT OUTER JOIN contact contact4_ ON sample0_.associatedContact_id=contact4_.id
LEFT OUTER JOIN District district5_ ON contact4_.district_id=district5_.id
LEFT OUTER JOIN cases case6_ ON contact4_.caze_id=case6_.id
LEFT OUTER JOIN District district7_ ON case6_.district_id=district7_.id
LEFT OUTER JOIN Person person8_ ON contact4_.person_id=person8_.id
LEFT OUTER JOIN EventParticipant eventparti9_ ON sample0_.associatedEventParticipant_id=eventparti9_.id
LEFT OUTER JOIN EVENTS event10_ ON eventparti9_.event_id=event10_.id
LEFT OUTER JOIN LOCATION location11_ ON event10_.eventLocation_id=location11_.id
LEFT OUTER JOIN District district12_ ON location11_.district_id=district12_.id
LEFT OUTER JOIN Person person13_ ON eventparti9_.person_id=person13_.id
LEFT OUTER JOIN Facility facility14_ ON sample0_.lab_id=facility14_.id
LEFT OUTER JOIN samples sample15_ ON sample0_.referredTo_id=sample15_.id
WHERE sample0_.deleted=?
ORDER BY sample0_.sampleDateTime DESC
```

</details>

It would help to get an analysis from the affected system ([how-to here](https://github.com/hzi-braunschweig/SORMAS-Project/blob/development/docs/TROUBLESHOOTING.md#run-analysis-of-a-sql-query-sormas-docker)).

**Comment by StefanKock** [comment]
We received an analysis that shows a cartesic explosion from ~8_000 lines to ~8_000_000 lines because of duplicate join to event participants: https://explain.dalibo.com/plan/U6t#plan/node/6

```sql
LEFT OUTER JOIN EventParticipant eventparti11_ ON sample0_.associatedEventParticipant_id=eventparti11_.id
LEFT OUTER JOIN EventParticipant eventperso15_ ON event12_.id=eventperso15_.event_id  -- additional
LEFT OUTER JOIN cases case16_ ON eventperso15_.resultingCase_id=case16_.id  -- additional
LEFT OUTER JOIN contact contacts17_ ON case16_.id=contacts17_.caze_id  -- additional
```

```bash
# Problem
EventParticipant -> Event -> EventParticipant -> Case -> Contact
11                  12       15                  16      17

# Probable Solution
EventParticipant -> Event
11                  12   
EventParticipant -> Case -> Contact
11                  16      17
```

<details>
<summary>Adapted SQL without duplicate EventParticipant join</summary>

`eventperso15_` is replaced by `eventparti11_` and commented out.

```sql
SELECT DISTINCT sample0_.uuid AS col_0_0_,
                case1_.epidNumber AS col_1_0_,
                sample0_.labSampleID AS col_2_0_,
                sample0_.sampleDateTime AS col_3_0_,
                sample0_.shipped AS col_4_0_,
                sample0_.shipmentDate AS col_5_0_,
                sample0_.received AS col_6_0_,
                sample0_.receivedDate AS col_7_0_,
                sample0_.sampleMaterial AS col_8_0_,
                sample0_.samplePurpose AS col_9_0_,
                sample0_.specimenCondition AS col_10_0_,
                facility19_.name AS col_11_0_,
                sample20_.uuid AS col_12_0_,
                sample0_.samplingReason AS col_13_0_,
                sample0_.samplingReasonDetails AS col_14_0_,
                case1_.uuid AS col_15_0_,
                person3_.firstName AS col_16_0_,
                person3_.lastName AS col_17_0_,
                contact5_.uuid AS col_18_0_,
                person10_.firstName AS col_19_0_,
                person10_.lastName AS col_20_0_,
                eventparti11_.uuid AS col_21_0_,
                person18_.firstName AS col_22_0_,
                person18_.lastName AS col_23_0_,
                CASE
                    WHEN case1_.id IS NOT NULL THEN case1_.disease
                    ELSE CASE
                             WHEN contact5_.id IS NOT NULL THEN contact5_.disease
                             ELSE event12_.disease
                         END
                END AS col_24_0_,
                CASE
                    WHEN case1_.id IS NOT NULL THEN case1_.diseaseDetails
                    ELSE CASE
                             WHEN contact5_.id IS NOT NULL THEN contact5_.diseaseDetails
                             ELSE event12_.diseaseDetails
                         END
                END AS col_25_0_,
                sample0_.pathogenTestResult AS col_26_0_,
                sample0_.additionalTestingRequested AS col_27_0_,
                EXISTS
  (SELECT additional21_.id
   FROM AdditionalTest additional21_
   WHERE sample0_.id=additional21_.sample_id) AS col_28_0_,
                       CASE
                           WHEN district2_.id IS NOT NULL THEN district2_.name
                           ELSE CASE
                                    WHEN district6_.id IS NOT NULL THEN district6_.name
                                    ELSE CASE
                                             WHEN district8_.id IS NOT NULL THEN district8_.name
                                             ELSE district14_.name
                                         END
                                END
                       END AS col_29_0_,
                       facility19_.uuid AS col_30_0_,

  (SELECT count(DISTINCT pathogente22_.id)
   FROM PathogenTest pathogente22_
   WHERE pathogente22_.sample_id=sample0_.id
     AND pathogente22_.deleted=FALSE) AS col_31_0_,
                       CASE
                           WHEN (sample0_.reportingUser_id IS NOT NULL)
                                AND sample0_.reportingUser_id=4468
                                OR 0=1
                                OR (case1_.reportingUser_id IS NOT NULL)
                                AND case1_.reportingUser_id=4468
                                OR case1_.responsibleRegion_id=138
                                OR case1_.region_id=138
                                OR (contact5_.reportingUser_id IS NOT NULL)
                                AND contact5_.reportingUser_id=4468
                                OR contact5_.region_id=138
                                OR case7_.responsibleRegion_id=138
                                OR case7_.region_id=138
                                OR (eventparti11_.reportingUser_id IS NOT NULL)
                                AND eventparti11_.reportingUser_id=4468
                                OR eventparti11_.region_id=138
                                OR location13_.region_id=138 THEN TRUE
                           ELSE FALSE
                       END AS col_32_0_,
                       CASE
                           WHEN (case1_.id IS NOT NULL)
                                AND ((case1_.reportingUser_id IS NOT NULL)
                                     AND case1_.reportingUser_id=4468
                                     OR case1_.responsibleRegion_id=138
                                     OR case1_.region_id=138) THEN TRUE
                           ELSE FALSE
                       END AS col_33_0_,
                       CASE
                           WHEN (contact5_.id IS NOT NULL)
                                AND ((contact5_.reportingUser_id IS NOT NULL)
                                     AND contact5_.reportingUser_id=4468
                                     OR contact5_.region_id=138
                                     OR case7_.responsibleRegion_id=138
                                     OR case7_.region_id=138
                                     OR (case7_.reportingUser_id IS NOT NULL)
                                     AND case7_.reportingUser_id=4468
                                     OR case7_.responsibleRegion_id=138
                                     OR case7_.region_id=138) THEN TRUE
                           ELSE FALSE
                       END AS col_34_0_,
                       CASE
                           WHEN (contact5_.id IS NOT NULL)
                                AND (case7_.id IS NOT NULL)
                                AND ((case7_.reportingUser_id IS NOT NULL)
                                     AND case7_.reportingUser_id=4468
                                     OR case7_.responsibleRegion_id=138
                                     OR case7_.region_id=138) THEN TRUE
                           ELSE FALSE
                       END AS col_35_0_,
                       CASE
                           WHEN (eventparti11_.id IS NOT NULL)
                                AND ((eventparti11_.reportingUser_id IS NOT NULL)
                                     AND eventparti11_.reportingUser_id=4468
                                     OR eventparti11_.region_id=138
                                     OR location13_.region_id=138) THEN TRUE
                           ELSE FALSE
                       END AS col_36_0_
FROM samples sample0_
LEFT OUTER JOIN cases case1_ ON sample0_.associatedCase_id=case1_.id
LEFT OUTER JOIN District district2_ ON case1_.district_id=district2_.id
LEFT OUTER JOIN Person person3_ ON case1_.person_id=person3_.id
LEFT OUTER JOIN contact contacts4_ ON case1_.id=contacts4_.caze_id
LEFT OUTER JOIN contact contact5_ ON sample0_.associatedContact_id=contact5_.id
LEFT OUTER JOIN District district6_ ON contact5_.district_id=district6_.id
LEFT OUTER JOIN cases case7_ ON contact5_.caze_id=case7_.id
LEFT OUTER JOIN District district8_ ON case7_.district_id=district8_.id
LEFT OUTER JOIN contact contacts9_ ON case7_.id=contacts9_.caze_id
LEFT OUTER JOIN Person person10_ ON contact5_.person_id=person10_.id
LEFT OUTER JOIN EventParticipant eventparti11_ ON sample0_.associatedEventParticipant_id=eventparti11_.id
LEFT OUTER JOIN EVENTS event12_ ON eventparti11_.event_id=event12_.id
LEFT OUTER JOIN LOCATION location13_ ON event12_.eventLocation_id=location13_.id
LEFT OUTER JOIN District district14_ ON location13_.district_id=district14_.id
-- LEFT OUTER JOIN EventParticipant eventperso15_ ON event12_.id=eventperso15_.event_id
LEFT OUTER JOIN cases case16_ ON eventparti11_.resultingCase_id=case16_.id
LEFT OUTER JOIN contact contacts17_ ON case16_.id=contacts17_.caze_id
LEFT OUTER JOIN Person person18_ ON eventparti11_.person_id=person18_.id
LEFT OUTER JOIN Facility facility19_ ON sample0_.lab_id=facility19_.id
LEFT OUTER JOIN samples sample20_ ON sample0_.referredTo_id=sample20_.id
WHERE ((event12_.id IS NOT NULL)
       AND (event12_.disease IS NULL)
       OR CASE
              WHEN case1_.id IS NOT NULL THEN case1_.disease
              WHEN contact5_.id IS NOT NULL THEN contact5_.disease
              ELSE event12_.disease
          END='CORONAVIRUS')
  AND ((case1_.region_id=138
        OR case1_.responsibleRegion_id=138
        OR (contacts4_.reportingUser_id='4468'
            OR contacts4_.contactOfficer_id='4468'
            OR contacts4_.region_id='138')
        AND (contacts4_.disease='CORONAVIRUS'
             OR contacts4_.disease IS NULL))
       AND case1_.disease='CORONAVIRUS'
       OR case1_.reportingUser_id=4468
       OR case1_.surveillanceOfficer_id=4468
       OR case1_.caseOfficer_id=4468
       OR (contact5_.reportingUser_id='4468'
           OR contact5_.contactOfficer_id='4468'
           OR contact5_.region_id='138')
       AND (contact5_.disease='CORONAVIRUS'
            OR contact5_.disease IS NULL)
       OR (case7_.region_id=138
           OR case7_.responsibleRegion_id=138
           OR (contacts9_.reportingUser_id='4468'
               OR contacts9_.contactOfficer_id='4468'
               OR contacts9_.region_id='138')
           AND (contacts9_.disease='CORONAVIRUS'
                OR contacts9_.disease IS NULL))
       AND case7_.disease='CORONAVIRUS'
       OR case7_.reportingUser_id=4468
       OR case7_.surveillanceOfficer_id=4468
       OR case7_.caseOfficer_id=4468
       OR (location13_.region_id='138'
           OR (case16_.region_id=138
               OR case16_.responsibleRegion_id=138
               OR (contacts17_.reportingUser_id='4468'
                   OR contacts17_.contactOfficer_id='4468'
                   OR contacts17_.region_id='138')
               AND (contacts17_.disease='CORONAVIRUS'
                    OR contacts17_.disease IS NULL))
           AND case16_.disease='CORONAVIRUS'
           OR case16_.reportingUser_id=4468
           OR case16_.surveillanceOfficer_id=4468
           OR case16_.caseOfficer_id=4468
           OR eventparti11_.region_id=138
           OR location13_.region_id='138'
           OR event12_.reportingUser_id='4468'
           OR event12_.responsibleUser_id='4468')
       AND (event12_.disease='CORONAVIRUS'
            OR event12_.disease IS NULL))
  AND (case1_.archived='f'
       OR case1_.archived IS NULL)
  AND sample0_.deleted='f'
ORDER BY sample0_.sampleDateTime DESC
LIMIT '100';
```

</details>

**Comment by StefanKock** [comment]
With the last given analysis I consider this ticket closed since #9779's improvement reduced the load time to 1% (1min to 300ms).
